Added namelist flags for previously uncontrolled output

Johannesfjeldsaa opened the pull request:
**Intention**: Enable capabilities to remove all outputs from the CAM-Oslo as opposed to previous tags e.g. noresm2_3_alpha03 where some outputs where outputted every time. 

**Method**: Introducing new namelist variables to control the history. The bellow table show which flag affects what output, further details on description available on betzy on path /cluster/projects/nn9560k/johannef/NorESM_workdir/cases-cam_output_reduction/README.md.
![image](https://github.com/user-attachments/assets/c4f26467-2b2f-4c01-a77a-f14da15e8648)

**Description of changes in OsloAero**: 
- src/oslo_aero_ndrop.F90: removed the history_amwg flag as it was not used in the file
- src/oslo_aero_ocean.F90: introduce *history_aerosol_forcing* flag to control odms output.
- src_cam/mo_chm_diags.F90: introduce *history_aerosol_base*, *history_aerosol_decomposed* and *history_gas* flag to control output of aerosol and gasphase aerosols. Some added inline comments for documentation added as well. Changes to the calculation of compound aerosol column burdens, they now include the cb of cloud water tracers. 
- src_cam/mo_gas_phase_chemdr.F90: introduce *history_aerosol_forcing* flag to control output of oxidant concentrations previously added for debugging. 


**How to test functionality/view my changes**:
- I have ran using `compset`=NF1850norbc and `resolution`=f19_f19. 
- Some changes are done to CAM as well. To run use the 'reduce_noresm2_3_default_output' branch in my fork of CAM https://github.com/Johannesfjeldsaa/CAM/tree/reduce_noresm2_3_default_output and point it to this version of OSLO_AERO in Externals_cam.cfg. Ofc then NorESM should point to my fork of CAM in it's Externals.cfg.
- I have ran tests with this setup with casedirectories available on betzy on path /cluster/projects/nn9560k/johannef/NorESM_workdir/cases-cam_output_reduction/ . Test 4-8 should cover the new capabilities, further descripitions is given in the README.md. A display of atmospheric history files are given in this notebook /cluster/projects/nn9560k/johannef/NorESM_workdir/cases-cam_output_reduction/look at output.ipynb. 

**Further work**:
- Merge CAM changes from https://github.com/Johannesfjeldsaa/CAM/tree/reduce_noresm2_3_default_output to https://github.com/NorESMhub/CAM/tree/noresm2_3_develop

Johannesfjeldsaa committed:
implemented flags history_aerosol_base, history_aerosol_decomposed and history_aerosol_gasphase to enable control of outputs via namelists.

Johannesfjeldsaa committed:
implemented history_aerosol_oxidant to enable controll of  outputs via nl

Johannesfjeldsaa committed:
Merge branch 'NorESMhub:noresm2_3_develop' into noresm2_3_develop

Johannesfjeldsaa committed:
implemented flags history_aerosol_base, history_aerosol_decomposed and history_aerosol_gasphase to enable control of outputs via namelists.

Johannesfjeldsaa committed:
remove definition of history_amwg as it is not used in oslo_aero_ndrop

Johannesfjeldsaa committed:
implemented flags history_aerosol_base, history_aerosol_decomposed and history_aerosol_gasphase to enable control of outputs via namelists.

Johannesfjeldsaa committed:
implemented flags history_aerosol_oxidant to enable control of outputs via namelists.

Johannesfjeldsaa committed:
making sure that H2O is outputted if history_aerosol_gasphase  is true

Johannesfjeldsaa commented:
@gold2718, @oyvindseland, @DirkOlivie @MichaelSchulzMETNO  - Could one of you review and test?

gold2718 commented:
@Johannesfjeldsaa, is this part of an effort to reduce output in NorESM2.3?
Are you planning on doing this, or something similar, for NorESM3?

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -116,7 +119,11 @@ subroutine chm_diags_inti
+                       history_dust_out = history_dust, &
+                       history_aerosol_base_out = history_aerosol_base, &              ! OSLO_AERO begin
+                       history_aerosol_decomposed_out = history_aerosol_decomposed, &
+                       history_aerosol_gasphase_out = history_aerosol_gasphase )       ! OSLO_AERO end
```
New control items should not be added to phys_getopts. Use protected public module variables instead.

Johannesfjeldsaa commented:
Yes, the first step is 2.3, as it is needed for dust PPE. We will then
bring this into the Noresm3 development branches.

tor. 13. mars 2025, 16:41 skrev goldy ***@***.***>:

> @Johannesfjeldsaa <https://github.com/Johannesfjeldsaa>, is this part of
> an effort to reduce output in NorESM2.3?
> Are you planning on doing this, or something similar, for NorESM3?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NorESMhub/OSLO_AERO/pull/51#issuecomment-2721702283>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXRQBABXSIGCOIEE75ZIFJD2UGRI7AVCNFSM6AAAAABY6A7YAKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRRG4YDEMRYGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: gold2718]*gold2718* left a comment (NorESMhub/OSLO_AERO#51)
> <https://github.com/NorESMhub/OSLO_AERO/pull/51#issuecomment-2721702283>
>
> @Johannesfjeldsaa <https://github.com/Johannesfjeldsaa>, is this part of
> an effort to reduce output in NorESM2.3?
> Are you planning on doing this, or something similar, for NorESM3?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NorESMhub/OSLO_AERO/pull/51#issuecomment-2721702283>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXRQBABXSIGCOIEE75ZIFJD2UGRI7AVCNFSM6AAAAABY6A7YAKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRRG4YDEMRYGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gold2718 commented:
Can you post your CAM branch as a (possibly draft) PR so I can compare the whole chain?

DirkOlivie reviewed (commented):
Seems fine to me.  Having odms under oxidants groups seems a bit strange.

MichaelSchulzMETNO commented:
I would rename the group history_aerosol_gasphase  to history_gas

oyvindseland commented:
I agree with Michael's point on name
Rename the group history_aerosol_gasphase to history_gas

Johannesfjeldsaa committed:
Including the column burden of cloud species into the compound aerosol cb. E.g.: in the compunded cb_DUST will now be the sum of DST_A3, DST_A3_OCW, DST_A2, DST_A2_OCW as opposed to previous sum which was DST_A3, DST_A2.

Johannesfjeldsaa committed:
renaming history_aerosol_gasphase to history_gas and history_aerosol_oxidant to history_aerosol_forcing

gold2718 commented on src/oslo_aero_ocean.F90:
```diff
@@ -182,8 +184,13 @@ subroutine oslo_aero_ocean_init()
             cycle_yr(m), fixed_ymd, fixed_tod, data_type(m) )
     enddo
+
+    call phys_getopts( history_aerosol_forcing_out = history_aerosol_forcing )
```
This should be a use statement at the top of the subroutine:
```suggestion
    use phys_control, only: history_aerosol_forcing
```

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -116,7 +119,11 @@ subroutine chm_diags_inti
+                       history_dust_out = history_dust, &
+                       history_aerosol_base_out = history_aerosol_base, &              ! OSLO_AERO begin
+                       history_aerosol_decomposed_out = history_aerosol_decomposed, &
+                       history_gas_out = history_gas )       ! OSLO_AERO end
```
These should be in a use statement at the top of the subroutine
```suggestion
                       history_dust_out = history_dust)
      use phys_control, only: history_aerosol_base, &              ! OSLO_AERO begin
                       history_aerosol_decomposed, &
                       history_gas        ! OSLO_AERO end
```

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +472,100 @@ subroutine chm_diags_inti
+            if ( history_aerosol_decomposed ) then
+               call add_default( spc_name, 1, ' ' )
+            endif
+         ! if it is not an aerosol species it is a gas species and we then require the history_gas flagg
```
Typo
```suggestion
         ! if it is not an aerosol species it is a gas species and we then require the history_gas flag
```

Johannesfjeldsaa replied:
will fix

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +472,100 @@ subroutine chm_diags_inti
+         cloudTracerIndex_direct = getCloudTracerIndexDirect(n)
+         if ( cloudTracerIndex_direct > 0 ) then
+            ! first the 3d fields, 
+            cloudTracerName(1:len(CloudTracerName))=" "
+            cloudTracerName = getCloudTracerName(n)
```
Is it necessary to initialize the string to blank?
```suggestion
            cloudTracerName = getCloudTracerName(n)
```
As stated in the Modern Fortran Programming book:
"If, however, the length of the result of the right-hand side is shorter than the length of the left-hand side, then the result is placed in the left-most part of the left-hand side and the rest is filled with blank characters."

Johannesfjeldsaa replied:
will remove the excess code

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +472,100 @@ subroutine chm_diags_inti
+            ! first the 3d fields, 
+            cloudTracerName(1:len(CloudTracerName))=" "
+            cloudTracerName = getCloudTracerName(n)
+            call addfld( trim(cloudTracerName), (/'lev'/), 'A','kg/kg', &
+               trim(cloudTracerName)//' in cloud water')
```
A consequence of putting the `addfld` call _inside_ the `cloudTracerIndex_direct > 0` conditional is that if you add the field manually (say to `fincl2` in user_nl_cam) the field will not be output and in fact, the model will crash during initialization. Is this what is desired (e.g., if the value is only computed  if `cloudTracerIndex_direct > 0`)?

gold2718 replied:
Is this really intentional? Are these quantities invalid if `cloudTracerIndex_direct <= 0`?

gold2718 replied:
I am still interested in understanding this.

gold2718 replied:
Due to code movement, this comment is no longer relevant, thanks for pointing that out.

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +472,100 @@ subroutine chm_diags_inti
+            if ( history_aerosol_decomposed ) then
+               call add_default('cb_'//trim(spc_name),1,' ' )
+            endif
+         ! else, if it is a gasphase the cb is included in the base so we require history_aerosol_base flagg
```
skrivefeil
```suggestion
         ! else, if it is a gasphase the cb is included in the base so we require history_aerosol_base flag
```

Johannesfjeldsaa replied:
will fix

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -657,104 +709,105 @@ subroutine chm_diags( lchnk, ncol, vmr, mmr, rxt_rates, invariants, depvel, depf
-       if ( any( tfy_species == m ) ) then
-          vmr_tfy(:ncol,:) = vmr_tfy(:ncol,:) +  wgt * vmr(:ncol,:,m)
-       endif
+   do m = 1,gas_pcnst
+
+   !...FOY (counting Fluorines, not chlorines or bromines)
+      if ( m == id_cfc12 .or. m == id_hcfc22 .or. m == id_cf2clbr .or. m == id_h1202 .or. m == id_hcfc142b &
+         .or. m == id_cof2 ) then
+         wgt = 2._r8
+      elseif ( m == id_cfc113 .or. m == id_cf3br ) then
+         wgt = 3._r8
+      elseif ( m == id_cfc114 .or. m == id_h2402 ) then
+         wgt = 4._r8
+      elseif ( m == id_cfc115 ) then
+         wgt = 5._r8
+      else
+         wgt = 1._r8
+      endif
+      if ( any( foy_species == m ) ) then
+         vmr_foy(:ncol,:) = vmr_foy(:ncol,:) +  wgt * vmr(:ncol,:,m)
+      endif
+      if ( any( tfy_species == m ) ) then
+         vmr_tfy(:ncol,:) = vmr_tfy(:ncol,:) +  wgt * vmr(:ncol,:,m)
+      endif
```
Is there a reason for the whitespace change in this region?

Johannesfjeldsaa replied:
"reflex code fix". Probably found the "non-default" whitespace annoying

gold2718 replied:
In that case, please make the indentation consistent, thanks!

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -657,104 +709,105 @@ subroutine chm_diags( lchnk, ncol, vmr, mmr, rxt_rates, invariants, depvel, depf
-       else
-          wgt = 1._r8
-       endif
+      if ( m == id_ch4 .or. m == id_n2o5 .or. m == id_cfc12 .or. m == id_cl2 .or. m == id_cl2o2 .or. m==id_h2o2  ) then
+         wgt = 2._r8
+      elseif (m == id_cfc114 .or. m == id_hcfc141b .or. m == id_h1202 .or. m == id_h2402 .or. m == id_ch2br2 ) then
+         wgt = 2._r8
+      elseif ( m == id_cfc11 .or. m == id_cfc113 .or. m == id_ch3ccl3 .or. m == id_chbr3 ) then
+         wgt = 3._r8
+      elseif ( m == id_ccl4 ) then
+         wgt = 4._r8
+      else
+         wgt = 1._r8
+      endif
```
Is there a reason for the whitespace change in this region (between here and line 810)?

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -767,20 +820,25 @@ subroutine chm_diags( lchnk, ncol, vmr, mmr, rxt_rates, invariants, depvel, depf
          endif
 
+         ! Add the column burden of the cloud tracer to the aerosol type cb
+         if (aerosolType(n) .gt. 0) then
```
Please use modern notation:
```suggestion
         if (aerosolType(n) > 0) then
```

Johannesfjeldsaa replied:
Sure, just reusing some old logical statements but will not use it anymore.

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -767,20 +820,25 @@ subroutine chm_diags( lchnk, ncol, vmr, mmr, rxt_rates, invariants, depvel, depf
-         ! Sum column burden per aerosol type
-         if(aerosolType(n) .gt. 0)then
+         ! Add the column burden and mass mixing ratio of the interstitial tracers to the aerosol type cb and mmr
+         if (aerosolType(n) .gt. 0) then
```
When editing, please convert to modern notation:
```suggestion
         if (aerosolType(n) > 0) then
```

Johannesfjeldsaa replied:
will fix

gold2718 commented on src_cam/mo_gas_phase_chemdr.F90:
```diff
@@ -29,6 +29,7 @@ module mo_gas_phase_chemdr
   integer :: ndx_cldfr, ndx_cmfdqr, ndx_nevapr, ndx_cldtop, ndx_prain
   integer :: ndx_h2so4
   ! OSLO_AERO begin
+  logical :: history_aerosol_forcing            ! history flagg for forcing diagnostic output
```
språkblanding
```suggestion
  logical :: history_aerosol_forcing            ! history flag for forcing diagnostic output
```

Johannesfjeldsaa replied:
will fix

gold2718 commented on src_cam/mo_gas_phase_chemdr.F90:
```diff
@@ -78,7 +79,10 @@ subroutine gas_phase_chemdr_inti()
-
+    
+    ! OSLO_AERO begin
+    call phys_getopts( history_aerosol_forcing_out = history_aerosol_forcing )
```
Convert this to a `use` statement at the top of the subroutine:
```suggestion
    use phys_control, only: history_aerosol_forcing
```

gold2718 reviewed (changes requested):
A few things need fixing (due to some changes needed in NorESMhub/CAM#198).

Johannesfjeldsaa committed:
Implementing PR review changes:
- oslo_aero_ocean.F90  and mo_gas_phase_chemdr.F90:  history_aerosol_forcing flag imported from phys_control module variables as opposed to via phys_getopts.
- mo_chm_diags.F90: replace .gt. with > in logic statements. Some spelling fixed. history_aerosol_base, history_aerosol_decomposed and history_gas flag imported from phys_control module variables as opposed to via phys_getopts. TODO - revert whitespace changes in code line 712-837

Johannesfjeldsaa commented:
@gold2718, @oyvindseland @MichaelSchulzMETNO @DirkOlivie 

For the compounded fields we have a choice of whether or not we want to have fields in terms of chemical species or the full aerosols. On example: do we give DDF of sulfate in terms of sulfur weight or sulfate weights that is either
SULFATE_DDF = SO4_A1DDF/3.06 + SO4_A2DDF/3.59 + ... + SO4_PR_OCWDDF/3.06
or 
SULFATE_DDF = SO4_A1DDF + SO4_A2DDF + ... + SO4_PR_OCWDDF

For emission/source this is very important as the expressions become more complicated AND fields like GS_SOA_LV/SV are used in the source expressions both for SO2 and POM. 

In short - should we have the unweighted fields (i.e. the aerosol mass) as outputable fields, if so, which keywords should govern what?

MichaelSchulzMETNO commented:
The CF convention suggests aerosol mass nor sulfur mass in output fields. Which means we should output the unweighted sum. NetCDF attributes need to be set also carefully, to not confuse the postprocessing. Attention is needed if two species are combined, eg so2+so4 - does that happen? maybe not...

oyvindseland commented:
The problem is when one calculates budgets, but I agree the CF convention is important here. I discussed it with Johannes and originally thought using S fields, but given your argument I think it is best to keep total mass in base, but perhaps treat S in decomp.

MichaelSchulzMETNO commented:
I think, when computing budgets, you need to postprocess anyway, to eg combine DMS and SO2 and SO4 etc. As long as the unit is correct, this should be no problem. Old scripts might have a problem, yes. / Why would you keep the S unit in decomposition ?

Johannesfjeldsaa commented:
We have investigated a little further: 
1. it is fairly simple technically to compute the full emission/source/load/sink/chemical loss at runtime by passing some minor amount of data out of a module using module level protected variables. This also instinuates that we don't need to pass individual fields which require post-processing but rather just give the similar field as you would from diagnostics directly from the model. 
2. If we want to be able to get the full overview of budgets from _history_aerosol_base_ we need to provide the emissions etc in terms of sulfur. There are three tracers/compounded aerosol species for which this is applicable; dms, SO2 and sulfates. In total providing **both** a CF-convesion output and a sulfur/budget specific output will require only ~10 2D fields extra. That way we meet multiple requirments _and_ are consitent across tracers/compounded aerosol species.

MichaelSchulzMETNO commented:
fine with me, would be nice to use CMIP6 variable names for the aerosol mass output, if there is a choice of variable names.

Johannesfjeldsaa committed:
changes to enable calculation of column burdens in terms of sulfur mass for SULFATE, dms and SO2.
- aero_model.F90
    - module level: use statement for subroutines in oslo_aero_share that registreres sulfur and SOA mass fractions for different tracers.
    - aero_model_register: call to the subroutines
- oslo_aero_share.F90
    - module level: two new public subroutines, and two new arrays to hold sulfur and SOA mass fractions for different tracers. Definition of new tracer indexes for gas tracers.
    - aero_register: add gas species call to cnst_get_ind
    - sulfur_mass_fraction registrer: assign sulfur mass fractions to sulfurMassFraction
    - soa_mass_fraction_register: assign SOA mass fractions to SOAMassFraction
- mo_chm_diags.F90
    - chm_diags_inti:  addfld and add_default calls for new fields; cb_SULFATE_S, cb_DMS_S, cb_SO2_S all with unit kg*S/m2
    - chm_diags: outfld calls for the new fields and calculation of them using cb(:) weighted with appropriate sulfurMassFraction

Johannesfjeldsaa committed:
changes to enable calculation of column burdens in terms of sulfur mass for SULFATE, dms and SO2.
- aero_model.F90
    - module level: use statement for subroutines in oslo_aero_share that registreres sulfur and SOA mass fractions for different tracers.
    - aero_model_register: call to the subroutines
- oslo_aero_share.F90
    - module level: two new public subroutines, and two new arrays to hold sulfur and SOA mass fractions for different tracers. Definition of new tracer indexes for gas tracers.
    - aero_register: add gas species call to cnst_get_ind
    - sulfur_mass_fraction registrer: assign sulfur mass fractions to sulfurMassFraction
    - soa_mass_fraction_register: assign SOA mass fractions to SOAMassFraction
- mo_chm_diags.F90
    - chm_diags_inti:  addfld and add_default calls for new fields; cb_SULFATE_S, cb_DMS_S, cb_SO2_S all with unit kg*S/m2
    - chm_diags: outfld calls for the new fields and calculation of them using cb(:) weighted with appropriate sulfurMassFraction

Johannesfjeldsaa committed:
Implementing source and emission fields for DMS, isoprene and monoterp.
- aero_model: sink fields introduced: sink_DMS, sink_DMS_S, sink_isoprene and sink_monoterp. These are outputted in history_aerosol_base
- chemistry.F90: emission fields introduced: emis_DMS, emis_DMS_S, emis_isoprene, emis_monoterp.

Johannesfjeldsaa committed:
changes to enable calculation of column burdens in terms of sulfur mass for SULFATE, dms and SO2.
- aero_model.F90
    - module level: use statement for subroutines in oslo_aero_share that registreres sulfur and SOA mass fractions for different tracers.
    - aero_model_register: call to the subroutines
- oslo_aero_share.F90
    - module level: two new public subroutines, and two new arrays to hold sulfur and SOA mass fractions for different tracers. Definition of new tracer indexes for gas tracers.
    - aero_register: add gas species call to cnst_get_ind
    - sulfur_mass_fraction registrer: assign sulfur mass fractions to sulfurMassFraction
    - soa_mass_fraction_register: assign SOA mass fractions to SOAMassFraction
- mo_chm_diags.F90
    - chm_diags_inti:  addfld and add_default calls for new fields; cb_SULFATE_S, cb_DMS_S, cb_SO2_S all with unit kg*S/m2
    - chm_diags: outfld calls for the new fields and calculation of them using cb(:) weighted with appropriate sulfurMassFraction

Johannesfjeldsaa committed:
Implementing source and emission fields for DMS, isoprene and monoterp.
- aero_model: sink fields introduced: sink_DMS, sink_DMS_S, sink_isoprene and sink_monoterp. These are outputted in history_aerosol_base
- chemistry.F90: emission fields introduced: emis_DMS, emis_DMS_S, emis_isoprene, emis_monoterp.

Johannesfjeldsaa committed:
changes to enable calculation of column burdens in terms of sulfur mass for SULFATE, dms and SO2.
- aero_model.F90
    - module level: use statement for subroutines in oslo_aero_share that registreres sulfur and SOA mass fractions for different tracers.
    - aero_model_register: call to the subroutines
- oslo_aero_share.F90
    - module level: two new public subroutines, and two new arrays to hold sulfur and SOA mass fractions for different tracers. Definition of new tracer indexes for gas tracers.
    - aero_register: add gas species call to cnst_get_ind
    - sulfur_mass_fraction registrer: assign sulfur mass fractions to sulfurMassFraction
    - soa_mass_fraction_register: assign SOA mass fractions to SOAMassFraction
- mo_chm_diags.F90
    - chm_diags_inti:  addfld and add_default calls for new fields; cb_SULFATE_S, cb_DMS_S, cb_SO2_S all with unit kg*S/m2
    - chm_diags: outfld calls for the new fields and calculation of them using cb(:) weighted with appropriate sulfurMassFraction

Johannesfjeldsaa committed:
added mo_extrfrc.F90 as a shadow as we are intending to collect CMXF outfld's from it

Johannesfjeldsaa committed:
first attempt to summation fields for dry and wet deposition

Johannesfjeldsaa committed:
started subroutine for writing out summation fields

Johannesfjeldsaa committed:
emission fields for DMS, isoprene and monoterp. New emission fields for SALT and DUST

Johannesfjeldsaa committed:
- Fix dry  deposition for compound aerosols: SULFATE, BC, OM, SALT and DUST. changes include:
    - BCDDF now exclude BC_AXDDF
    - There is now only a weighted version for sulfur: SULFUR_SDDF
- Fix wet deposition for compound aerosols: BC, OM, SALT and DUST, changes include:
    - BCSFWET now exclude BC_AXDDF
    - no weigthed version for these compounded aerosol types
- Make compunded sulfur SFWET field a module level public protected variable so that wet deposition field can be calculated INCLUDING WD_A field in tphysbc

Johannesfjeldsaa committed:
- Fix dry  deposition for compound aerosols: SULFATE, BC, OM, SALT and DUST. changes include:
    - BCDDF now exclude BC_AXDDF
    - There is now only a weighted version for sulfur: SULFUR_SDDF
- Fix wet deposition for compound aerosols: BC, OM, SALT and DUST, changes include:
    - BCSFWET now exclude BC_AXDDF
    - no weigthed version for these compounded aerosol types
- Make compunded sulfur SFWET field a module level public protected variable so that wet deposition field can be calculated INCLUDING WD_A field in tphysbc

Johannesfjeldsaa committed:
remember to import things you are using ...

Johannesfjeldsaa committed:
Some confusion wrt BC_AX: it should be included in both the cb and for deposition fields
- Allow BC_AXDDF and BCAX_SFWET to contribute to their respective deposition fields.

Johannesfjeldsaa committed:
rename summation fields for deposition from DDF/SFWET to dry_/wet_. E.g:
SULFURDDF -> dry_SULFATE

Johannesfjeldsaa committed:
Summation fields across modules:
mo_extfrc.F90 - we collect CMXF fields which we use for emission fields.

mo_chm_diags.F90 - collect DF_SO2, WD_A_SO2 and WD_A_H2SO4 for use in sink field for SO2 and wet deposition field for SULFATE

chemistry.F90
- collect SF variables for emission fields.
- output of emis_SALT and emis_DUST
- initialize summation fields via summation_fields_init()
- summation_fields_writeout_bc and _ac subroutines which calcluates summation fields using data from across modules. _ac will only handle wet_SULFATE for now, is not called from enywhere yet as pbuf variable is needed.

oslo_aero_share.F90
- sulfur mass fraction registry, some changes to keep track of msa as sulfur
- SOA yield registry, split into reals instead of real(:) as there was only applicable for the terpens.

aero_model.F90
- export GS and AQ fields for summation fields.

Johannesfjeldsaa committed:
Summation fields across modules:
mo_extfrc.F90 - we collect CMXF fields which we use for emission fields.

mo_chm_diags.F90 - collect DF_SO2, WD_A_SO2 and WD_A_H2SO4 for use in sink field for SO2 and wet deposition field for SULFATE

chemistry.F90
- collect SF variables for emission fields.
- output of emis_SALT and emis_DUST
- initialize summation fields via summation_fields_init()
- summation_fields_writeout_bc and _ac subroutines which calcluates summation fields using data from across modules. _ac will only handle wet_SULFATE for now, is not called from enywhere yet as pbuf variable is needed.

oslo_aero_share.F90
- sulfur mass fraction registry, some changes to keep track of msa as sulfur
- SOA yield registry, split into reals instead of real(:) as there was only applicable for the terpens.

aero_model.F90
- export GS and AQ fields for summation fields.

Johannesfjeldsaa committed:
aero_model.F90:
- whitespace refactoring
- include call to oslo_aero_depos_register
- refactor from soa_mass_fraction_register() to soa_yield_register call
- correct error msg and endrun call for array allocations in aero_model_init

oslo_aero_depos.F90:
- whitespace refactoring
- define oslo_aero_depos_register subroutine
- introduce a pbuf variable at idx_wd_a_h2so4
- rename SFWET->wet_ and DDF->dry_ for the compound aerosols

oslo_aero_share.F90:
- whitespace refactoring
- refactor from soa_mass_fraction_register() to soa_yield_register call

chemistry.F90:
- whitespace refactoring
-  correct error msg and endrun call for array allocations in chem_init
- neu_wetdep_tend call includes pbuf argument
- additional summation field history

mo_chm_diags.F90:
- whitespace refactoring
- het_diags call includes pbuf argument
- correct error msg and endrun call for array allocations in chem_diags_inti
- fix bug in summation column burdens for BC. A logic test in line 883 was missing resulting in a double counting of one of the itersitial tracers.
- het_diags pbuf passing of wd_a_h2so4

mo_neu_wetdep.F90:
- pbuf passing of wd_a_h2so4
- module variable passing of wd_a_so2

TODO:
- OSLO_AERO_DEPOS:
     - remove wd_a_h2so4_debug history. It is in temporary to allow for double checking the summation fields.
- chemistry:
     - remove history for intermediate calculations in summeation_fides_init and writeout.

Johannesfjeldsaa committed:
wrong sign for two of the fields in the calculation of SO2 sink

Johannesfjeldsaa committed:
Moving the wet_SO2(_S) initialisation to mo_chm_diags from chemistry and outfld call to mo_neu_wetdep from chemistry,

Johannesfjeldsaa committed:
change to string passed to cnst_get_ind to pass the correct arrays to WD_A_SO2 and WD_A_H2SO4.

gold2718 commented on src/aero_model.F90:
```diff
@@ -1,1000 +1,1159 @@
+   use oslo_aero_ocean,          only: oslo_aero_ocean_init, oslo_aero_dms_emis
+   use oslo_aero_share,          only: getNumberofTracersInMode, getCloudTracerIndexDirect, getCloudTracerName
+   use oslo_aero_share,          only: getCloudTracerName, getTracerIndex, aero_register
+   use oslo_aero_share,          only: sulfur_mass_fraction_register, soa_yield_register
```
This line should be down in `aero_model_register`.

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -37,11 +39,13 @@ module oslo_aero_depos
   public :: oslo_aero_depos_wet ! wet deposition
   public :: oslo_aero_wetdep_init
 
+
```
Is extra blank line needed?

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -92,19 +96,30 @@ module oslo_aero_depos
 
     ! local variables
-    integer            :: m, l, i
+    integer            :: m, l, i, n
```
I know you didn't start it but single letter variables are not good practice (easy to get the wrong one, impossible to search for, etc.). Can you help break the mold here by using a meaningful variable name (e.g., `atype_ind`)?

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -277,12 +337,15 @@ subroutine oslo_aero_depos_dry  (lchnk, ncol, psetcols, &
     integer :: lphase                    ! index for interstitial / cloudborne aerosol
     integer :: lspec                     ! index for aerosol number / chem-mass / water-mass
     integer :: m                         ! aerosol mode index
+    integer :: n                         ! aerosol type index
```
Another plea for useful variable names here.

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -564,6 +651,7 @@ subroutine oslo_aero_depos_wet ( lchnk, ncol, psetcols, pmid, pdel, q, t, &
 
     ! Local variables
     integer  :: m                             ! tracer index
+    integer  :: n                             ! aerosol type index
```
Another opportunity to fight back against single-letter variables.

gold2718 commented on src/oslo_aero_depos.F90:
Is one blank line enough?

gold2718 commented on src/oslo_aero_depos.F90:
Props for limiting output to the atm.log file but do we need this in the log file every time? The printed value is only from one task, what is it telling us?

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -1475,6 +1609,7 @@ subroutine wetdepa_v2(                                    &
     !-----------------------------------------------------------------------
 
     real(r8), intent(in) ::&
+
```
Not a great place for a blank line IMHO.

gold2718 commented on src/oslo_aero_share.F90:
```diff
@@ -1,1084 +1,1146 @@
+   integer, private :: qqcw(pcnst)=-1 ! Remaps modal_aero indices into pbuf
+
+
 
```
Do we need



three blank lines?

gold2718 commented on src/oslo_aero_share.F90:
```diff
@@ -1,1084 +1,1146 @@
+      sulfurMassFraction(l_so2) = 1.0_r8/1.998_r8
+      ! for sulfates the sulfur mass fraction is variable depending on the
+      ! sulfate compound. For sulfate produced by aqueous-phase chemistry
+      ! the sulfur mass fraction is assumed ~1/3.59 since ...
```
These `...` comments have me on the edge of my seat! To be continued? Coming next season?
Also applies to 509, 514, 517, 530, and 532.

gold2718 replied:
I think I understand what you are trying to convey here but I think a more standard approach (in English anyway) is:
```suggestion
      ! the sulfur mass fraction is assumed ~1/3.59 since:
```

gold2718 commented on src_cam/chemistry.F90:
```diff
@@ -1,1513 +1,2018 @@
+      ! local vars
+
+      integer :: lchnk, ncol
+      integer :: i, m, n, n2
```
1. `n2` is not much better than just `n` (if at all).
2. Is it even used in this routine?

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -117,7 +132,21 @@ subroutine chm_diags_inti
+    allocate( DF_SO2(pcols, begchunk:endchunk), stat=astat )
+    if( astat/= 0 ) then
+      write(iulog,*) 'chm_diags_inti: failed to allocate DF_SO2 array; error = ',astat
+      call endrun('chm_diags_inti: failed to allocate DF_SO2 array')
```
You don't have to do this but you can simplify these error messages (using `use string_utils, only: int2str`):
```suggestion
      call endrun('chm_diags_inti: failed to allocate DF_SO2 array; error = '//int2str(astat))
```

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +508,120 @@ subroutine chm_diags_inti
+               call add_default( trim(cloudTracerName), 1, ' ' )
+            endif
+
+            ! and then then add column burden of cloud tracers
```
What comment is this a followup to?

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +508,120 @@ subroutine chm_diags_inti
+         call addfld('cb_'//trim(spc_name),horiz_only, 'A', 'kg/m2', &
+            'cb_'//trim(spc_name)//' in column')
+
+         ! ... and if it is so2 or dms we add the column burden of the sulfur
```
It's really hard to follow these widely-separated comments (not that I have a helpful suggestion on how to fix it).

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -463,57 +508,120 @@ subroutine chm_diags_inti
+   ! OSLO_AERO begin
+   ! iterate over the compounded aerosol types
    do n=1,N_AEROSOL_TYPES
+
```
Any particular reason for this blank line?

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -767,20 +881,45 @@ subroutine chm_diags( lchnk, ncol, vmr, mmr, rxt_rates, invariants, depvel, depf
          cb(:ncol) = sum(mass_tmp(:ncol,:),2)
          call outfld(trim('cb_'//trim(spc_name)), cb, pcols, lchnk)
+         if ( n == l_so2 .or. n == l_dms ) then
+            call outfld(trim('cb_'//trim(spc_name)//'_S'), ( cb(:ncol) * sulfurMassFraction(n) ), pcols, lchnk)
```
This is a big error that will output the wrong data! I will look but we need to make sure we stomp out all `ncol` vs. `pcols` mismatches. This one should be:
```suggestion
            call outfld(trim('cb_'//trim(spc_name)//'_S'), ( cb(:) * sulfurMassFraction(n) ), pcols, lchnk)
```
While you could use `ncol` in both places, it is slightly less efficient:
```suggestion
            call outfld(trim('cb_'//trim(spc_name)//'_S'), ( cb(:ncol) * sulfurMassFraction(n) ), ncol, lchnk)
```

gold2718 commented on src_cam/mo_extfrc.F90:
```diff
@@ -0,0 +1,463 @@
+ 
+    call phys_getopts( &
+         history_aerosol_out = history_aerosol, &
+         history_chemistry_out = history_chemistry, & 
```
No spaces at the end of the line please.

gold2718 commented on src_cam/mo_extfrc.F90:
```diff
@@ -0,0 +1,463 @@
+          endif
+       endif
+    enddo
+
+    
+
+
```
Do we really need all these blank lines? The original has only one here.

gold2718 commented on src_cam/mo_neu_wetdep.F90:
```diff
@@ -282,8 +310,11 @@ subroutine neu_wetdep_tend(lchnk,ncol,mmr,pmid,pdel,zint,tfld,delt, &
+  real(r8)    :: wrk_wd(pcols)
+  logical     :: history_aerosol
+  integer     :: l_aero
+  type(physics_buffer_desc),  pointer :: pbuf(:)
```
This should really be up with the rest of the dummy arguments, not under local variables.
While you are at it, you can remove line 267 (`implicit none`), it is redundant.

gold2718 commented on src_cam/mo_neu_wetdep.F90:
```diff
@@ -485,17 +519,38 @@ subroutine neu_wetdep_tend(lchnk,ncol,mmr,pmid,pdel,zint,tfld,delt, &
+
+      if ( l_aero == l_so2 ) then
+
+        call outfld('wet_SO2', wrk_wd(:ncol), ncol, lchnk)
+        call outfld('wet_SO2_S', ( wrk_wd(:ncol) * sulfurMassFraction(l_so2) ), ncol, lchnk)
```
Where are the `addfld` calls for these? They are usually (i.e., should be) in the same module as the `outfld` calls.

gold2718 reviewed (changes requested):
Much better but there is still one unresolved question from the first round and some items that need to be fixed (along with several suggestions).

Johannesfjeldsaa committed:
fix PR requests

Johannesfjeldsaa committed:
Added history_aerosol_debug_output flag to ensure that summation field controll calculations can be ran, but only when needed.

gold2718 commented on src_cam/mo_chm_diags.F90:
```diff
@@ -392,8 +422,22 @@ subroutine chm_diags_inti
        call addfld( wetdep_name_area(m), horiz_only, 'A', 'kg/m2/s ', spc_name//' wet deposition' )
 
+       call cnst_get_ind(spc_name, l_aero, abort=.false.)
+       if ( l_aero == l_so2 ) then
+
+         call addfld ('wet_SO2', horiz_only, 'A',  'kg/m2/s',   &
+            'SO2 wet deposition flux at surface.')
+         call addfld ('wet_SO2_S', horiz_only, 'A',  'kg*S/m2/s',   &
+            'SO2 wet deposition flux at surface, sulfur mass only.')
+
+         if ( history_aerosol_base ) then
+            call add_default( 'wet_SO2', 1, ' ' )
+            call add_default( 'wet_SO2_S', 1, ' ' )
+         end if
+       end if
```
Please move this to `neu_wetdep_init` in mo_neu_wetdep.F90

gold2718 reviewed (changes requested):
Great cleanup, thanks especially for all the improved variable names and improved `endrun` calls.
I have a couple of small comments and there are a couple of unresolved questions from before.

Johannesfjeldsaa committed:
PR requested changes and changing sign of wet deposition fields so that it is positive as opposed to negative.

gold2718 reviewed (changes requested):
Lots of good improvements, thanks. There is still one open question from the 24. March review.

Johannesfjeldsaa commented:
> Lots of good improvements, thanks. There is still one open question from the 24. March review.

@oyvindseland will account for this.

oyvindseland commented:
This is basically old code Johannes just changed the positions I think.

The critical line  is actually  if (n > 0) then 
If n is nonzero then getCloudTracerIndexDirect(n) must be defined.

gold2718 reviewed (approved):
:+1:

Ovewh commented:
There are still some instances where the history_aerosol boolean is used to enable some calculations and not only adding a certain diagnostic to namelist default. If these are expensive calculations we might not want to calculate this by default? Maybe it should be controlled by another namelist parameter, that only enables this calculations and then have `history_aerosol` to enable add_default calls. 
e.g. [oslo_aero_coag.F90](https://github.com/Johannesfjeldsaa/OSLO_AERO/blob/3cdbbe4a73189d350f26b41b14ea7faf45226e8a/src/oslo_aero_coag.F90#L544) and [oslo_aero_coagtend.F90](https://github.com/Johannesfjeldsaa/OSLO_AERO/blob/3cdbbe4a73189d350f26b41b14ea7faf45226e8a/src/oslo_aero_condtend.F90#L542)
```fortran
    if(history_aerosol)then
       coltend(:ncol,:) = 0.0_r8
       do i=1,gas_pcnst
          !Check if species contributes to coagulation
          if(lifeCycleReceiver(i) .gt. 0)then
             !Loss from the donor specie
             tracer_coltend(:ncol) = sum(totalLoss(:ncol, :,i)*pdel(:ncol,:),2)/gravit*delt_inverse
             coltend(:ncol,i) = coltend(:ncol,i) - tracer_coltend(:ncol) !negative, loss for donor
             coltend(:ncol,lifeCycleReceiver(i)) = coltend(:ncol,lifeCycleReceiver(i)) + tracer_coltend(:ncol)
          endif
       end do
       do i=1,gas_pcnst
          if(lifeCycleReceiver(i) .gt. 0)then
             long_name= trim(solsym(i))//"coagTend"
             call outfld(long_name, coltend(:ncol,i), ncol, lchnk)
             long_name= trim(solsym(lifeCycleReceiver(i)))//"coagTend"
             call outfld(long_name, coltend(:ncol,lifeCycleReceiver(i)),ncol,lchnk)
          end if
       end do
    endif

```

Johannesfjeldsaa committed:
remove wrapping of outfld calls and intermediated calculations by history_aerosol, i.e. allowing to output only using fincl

Johannesfjeldsaa committed:
remove wrapping of outfld calls and intermediated calculations by history_aerosol, i.e. allowing to output only using fincl

Johannesfjeldsaa commented:
> There are still some instances where the history_aerosol boolean is used to enable some calculations and not only adding a certain diagnostic to namelist default. If these are expensive calculations we might not want to calculate this by default? Maybe it should be controlled by another namelist parameter, that only enables this calculations and then have `history_aerosol` to enable add_default calls. e.g. [oslo_aero_coag.F90](https://github.com/Johannesfjeldsaa/OSLO_AERO/blob/3cdbbe4a73189d350f26b41b14ea7faf45226e8a/src/oslo_aero_coag.F90#L544) and [oslo_aero_coagtend.F90](https://github.com/Johannesfjeldsaa/OSLO_AERO/blob/3cdbbe4a73189d350f26b41b14ea7faf45226e8a/src/oslo_aero_condtend.F90#L542)
> 
> ```fortran
>     if(history_aerosol)then
>        coltend(:ncol,:) = 0.0_r8
>        do i=1,gas_pcnst
>           !Check if species contributes to coagulation
>           if(lifeCycleReceiver(i) .gt. 0)then
>              !Loss from the donor specie
>              tracer_coltend(:ncol) = sum(totalLoss(:ncol, :,i)*pdel(:ncol,:),2)/gravit*delt_inverse
>              coltend(:ncol,i) = coltend(:ncol,i) - tracer_coltend(:ncol) !negative, loss for donor
>              coltend(:ncol,lifeCycleReceiver(i)) = coltend(:ncol,lifeCycleReceiver(i)) + tracer_coltend(:ncol)
>           endif
>        end do
>        do i=1,gas_pcnst
>           if(lifeCycleReceiver(i) .gt. 0)then
>              long_name= trim(solsym(i))//"coagTend"
>              call outfld(long_name, coltend(:ncol,i), ncol, lchnk)
>              long_name= trim(solsym(lifeCycleReceiver(i)))//"coagTend"
>              call outfld(long_name, coltend(:ncol,lifeCycleReceiver(i)),ncol,lchnk)
>           end if
>        end do
>     endif
> ```

I have removed the ones I found in oslo_aero but have not looked elsewhere in the model. Thank you for pointing it out.

oyvindseland commented:
I agree that it should be implemented in a clean way now, i.e only use the history_flag outside add_default there are probably enough additional calculations that additional flags should be added to save some cpu time. The detailed budget keeping in Oslo-Aero is not the main reason why oslo-aero is more expensive than mam but it may contribute. This can at any rate wait until after the physics dead-line.
I think  sometimes overdo this though. For some reason the compound column burden in mam is only written out for daylight hours.

oyvindseland reviewed (approved):
I think the iteration has solved the problems
